### PR TITLE
Fix strip_paths_in_code() for Python 3.8

### PR DESCRIPTION
**Pull request by vstinner**
strip_paths_in_code() now uses CodeType.replace() if available to
support the new CodeType construction (PEP 570: Positional-Only
Parameters).

Co-Authored-By: Pablo Galindo Pablogsal@gmail.com

### CODE REVIEW

**Comment by lkollar** [comment]
Looks like the build on Python 3.5 failed due to a network issue:
`curl: (7) Failed to connect to whereami.travis-ci.com port 443: Connection timed out`

@vstinner could you force push to re-trigger it?

**Comment by vstinner** [comment]
> Looks like the build on Python 3.5 failed due to a network issue: ...

Sorry, which job? The Travis CI shows me green jobs https://travis-ci.org/pyinstaller/pyinstaller/builds/587485729 except for nightly which failed on building numpy, compilation error on "__Pyx_PyCode_New" from Cython. The numpy C code generated by Cython should be regenerated for Python 3.8 and newer.

**Comment by vstinner** [comment]
> Sorry, which job? The Travis CI shows me green jobs https://travis-ci.org/pyinstaller/pyinstaller/builds/587485729 except for nightly which failed on building numpy, compilation error on "__Pyx_PyCode_New" from Cython. The numpy C code generated by Cython should be regenerated for Python 3.8 and newer.

Oh. I didn't notice that the are "Libraries" checks, I forgot to scroll. Well, I'm not a PyInstaller developer, I'm not allowed to retrigger a single job.

I could close/reopen my PR to retrigger all CIs, but it seems like the CIs are unstable so I'm not sure what to do...

@lkollar: You are not allowed to retrigger a job on Travis CI once you are logged in Travis CI using GitHub authentication?

**Comment by lkollar** [comment]
@vstinner sorry my bad, I didn't realise with triage rights I could restart the build. Running it again now.

**Comment by vstinner** [comment]
@lkollar: CI tests now pass. Is there anything else to do to merge this fix? It's needed to support Python 3.8: see issue #4311.

**Comment by lkollar** [comment]
It would be great to merge this but I only have triage rights. We need a maintainer who can land it.

**Comment by vstinner** [comment]
> It would be great to merge this but I only have triage rights. We need a maintainer who can land it.

Is there a way to ping a maintainer or give you the maintainer permission? :-)

**Comment by lkollar** [comment]
Pinging @htgoebel @bjones1 

I can confirm that with the latest 3.8 CPython build and @vstinner's patch this works:

```
$ cat test.py
import sys
print(sys.version)
$ pyinstaller test.py
91 INFO: PyInstaller: 4.0.dev0+g82cf4a62.mod
91 INFO: Python: 3.9.0a0
99 INFO: Platform: macOS-10.14.6-x86_64-i386-64bit
<omitted for brevity>
$ ./test/dist/test/test
3.9.0a0 (heads/master:58498bc, Sep 29 2019, 19:01:55)
[Clang 10.0.1 (clang-1001.0.46.4)]
```

**Comment by bjones1** [comment]
@vstinner, thanks for your contributions. Would you create the file `news\4440.bugfix.rst`, which should briefly explain the changes you made. In particular, reference PEP 570, and if you can find any links/docs to `co_posonlyargcount` and/or `co_kwonlyargcount`, that would be helpful. The file's format is reST, so a typical link would be:

```
`PEP 570 <https://www.python.org/dev/peps/pep-0570/>`_
```

**Comment by vstinner** [comment]
> Would you create the file news\4440.bugfix.rst, ...

Done. I'm not sure if I should elaborate the technical details or not. Does it look good to you?

**Comment by bjones1** [comment]
This looks fine. Thanks!

**Comment by vstinner** [comment]
Thanks.

By the way, you may consider to use the new PyConfig API that I wrote for Python 3.8 to use configuration customization when embedding/using Python in C:
* https://docs.python.org/dev/c-api/init_config.html
* https://www.python.org/dev/peps/pep-0587/

**Comment by vstinner** [comment]
Ah, and FYI I wrote this PR to help Pablo (co-author of the PR) who had an issue with PyInstaller on Python 3.8. My PEP 587 implementation introduced some bugs in the "path configuration" (sys.path & friends).

By the way, Python 3.8 now dumps the path configuration if importing the Python codec of the filesystem encoding fails. It should help to debug path config issues ;-) https://twitter.com/VictorStinner/status/1176475423301148672

**Comment by bjones1** [comment]
The PyConfig API looks neat! I assuming it's only available starting in Python 3.8, correct?

**Comment by vstinner** [comment]
> The PyConfig API looks neat! I assuming it's only available starting in Python 3.8, correct?

It's new in Python 3.8.

**Comment by bjones1** [comment]
Since we support all current Python versions in a single bootloader, we can't use that until Python 3.7 is EOL, which will be a while...

**Comment by vstinner** [comment]
> Since we support all current Python versions in a single bootloader, we can't use that until Python 3.7 is EOL, which will be a while...

You might have one code path for Python 3.8 and newer, another one for older versions. Well, it's really up to you ;-)

PyConfig is supposed to be way more reliable at Python update: PyConfig is better defined. For example, you have to choose the starting configuration: "Python" configuration or "isolated" configuration.

It's not easy to discover all Python configuration options. They are scattered around the code, and we only started to document them recently!
https://docs.python.org/dev/c-api/init.html#global-configuration-variables

Some options are set by global configuration variables, some others by function calls.

Example from PyInstaller:

```
    /*
     * Startup flags - default values. 1 means enabled, 0 disabled.
     */
    /* Suppress 'import site'. */
    *PI_Py_NoSiteFlag = 1;
    /* Needed by getpath.c from Python. */
    *PI_Py_FrozenFlag = 1;
    /* Suppress writing bytecode files (*.py[co]) */
    *PI_Py_DontWriteBytecodeFlag = 1;
    /* Do not try to find any packages in user's site directory. */
    *PI_Py_NoUserSiteDirectory = 1;
    /* This flag ensures PYTHONPATH and PYTHONHOME are ignored by Python. */
    *PI_Py_IgnoreEnvironmentFlag = 1;
    /* Disalbe verbose imports by default. */
    *PI_Py_VerboseFlag = 0;
```

PyConfig makes it easier:

```
PyConfig_InitIsolatedConfig(&config)
```

It's also difficult to handle encodings "properly": the PR #4441 is a good example. PyInstaller:

```
    static wchar_t pyhome_w[PATH_MAX + 1];
    static wchar_t progname_w[PATH_MAX + 1];
...
        /* Decode using current locale */
        if (!pyi_locale_char2wchar(pyhome_w, status->mainpath, PATH_MAX)) {
            FATALERROR("Failed to convert pyhome to wchar_t\n");
            return -1;
        }
        VS("LOADER: sys.prefix is %s\n", status->mainpath);
        PI_Py_SetPythonHome(pyhome_w);
...
        /* Decode using current locale */
        if (!pyi_locale_char2wchar(progname_w, status->archivename, PATH_MAX)) {
            FATALERROR("Failed to convert progname to wchar_t\n");
            return -1;
        }
        /* In Python 3 Py_SetProgramName() should be called before Py_SetPath(). */
        PI_Py_SetProgramName(progname_w);
```

PyConfig:

```
PyStatus status;

status = PyConfig_SetBytesString(&config, &config, &config.home, status->mainpath);
if (PyStatus_Exception(status)) { ... handle the error ... }

status = PyConfig_SetBytesString(&config, &config, &config.program_name, status->archivename);
if (PyStatus_Exception(status)) { ... handle the error ... }
```

PyConfig gives a fine control on how Python is initialized. Example from PyInstaller:

```
    if (unbuffered) {
#ifdef _WIN32
        _setmode(fileno(stdin), _O_BINARY);
        _setmode(fileno(stdout), _O_BINARY);
#endif
        fflush(stdout);
        fflush(stderr);

        setbuf(stdin, (char *)NULL);
        setbuf(stdout, (char *)NULL);
        setbuf(stderr, (char *)NULL);
    }
```

Same with PyConfig, no need to acess low-level setbuf() functions or `#ifdef _WIN32`:

```
config.buffered_stdio = 0;
config.configure_c_stdio = 1;
```

* https://docs.python.org/dev/c-api/init_config.html#c.PyConfig.configure_c_stdio
* https://docs.python.org/dev/c-api/init_config.html#c.PyConfig.buffered_stdio

Well, I didn't look at the code careful, maybe PyConfig API lacks some helper for more complex tasks like the path configuration (sys.path & friends).

Note: PyConfig avoids double negation and has better names, it's simpler to read & understand ;-)

File: PyInstaller/building/utils.py [hunk 1]
```diff
@@ -657,8 +657,10 @@ def strip_paths_in_code(co, new_filename=None):
         for const_co in co.co_consts
     )
 
+    if hasattr(co, 'co_posonlyargcount'):
+        return co.replace(co_consts=consts)
```
Comment (htgoebel) [inline PyInstaller/building/utils.py]: The changelog entry says: "use the new ``CodeType.replace()`` method if available." But the code does not test for `.replace`, but for `co_posonlyargcount`. While the later might be an indicator for the former, it is a different test.

Anyway: When using `if is_py38` as a condition, we could spot this as a place where the othercode could be removed in some years when we drop support for Python 3.7.

Comment (htgoebel) [reply PyInstaller/building/utils.py]: (I know `is_py38` is not yet here. It should get into live on te python-3.8 branch.)

**Comment by htgoebel** [comment]
@vstinner Please open a new issue for using the new PyConfig style. THis very issue is about `strip_paths_in_code()`. Not poluting other issues helps us keeping track. Many Thanks.

File: PyInstaller/building/utils.py (hunk 1 above)
Comment (vstinner) [reply PyInstaller/building/utils.py]: This PR is already merged. You can replace hasattr(co, 'co_posonlyargcount') with hasattr(co, 'replace') if you want ;-) In practice, it's basically the same.

**Comment by htgoebel** [comment]
Sorry to say, bat this was a pure pull-request: Not only the new code checks for `co_posonlyargcount` instead of `replace` (see https://github.com/pyinstaller/pyinstaller/pull/4440#discussion_r344761636), and the commit message does not comply to our conventions. Much more important:

**It broke the relevant part of this method**: replacing the *filename*.

I just fixed it in f1813d5543c6417c5a7e9481ee288532249cc305.

Leasons learned: For new versions of Python, don't implement single, random fixes. Instead first set up a branch, activate CI testing, then start coding.

**Comment by vstinner** [comment]
Thanks for the fix.

> Leasons learned: For new versions of Python, don't implement single, random fixes. Instead first set up a branch, activate CI testing, then start coding.

It seems like an unit test is missing :-) At least, with the incorrect fix, I managed to build a binary :-) Without the fix, it failed with a Python exception.

**Pull request merged**